### PR TITLE
Issue #518: Create `git machete help config` topic for config keys and env vars

### LINKED ISSUE #518: Create a new help topic (not command!) `git machete help config`
To group all the configs & env vars since there's already quite a few of them:

* `GIT_MACHETE_REBASE_OPTS`
* `machete.github.{remote,organization,repository}` (in progress on @amalota's PR #487)
* `machete.overrideForkPoint.<branch>.{to,whileDescendantOf}`
* `machete.status.extraSpaceBeforeBranchName` (TBD on @amalota's PR #511)
* `machete.worktree.useTopLevelMacheteFile` (??? &mdash; TBD for @1wsx10 issue #517)
* (check if I haven't forgotten anything)

### PR DESCRIPTION
**Pull request by amalota**

### CODE REVIEW

**Comment by amalota** [comment]
Result:
image

File: git_machete/cli.py [hunk 1]
```diff
@@ -31,7 +31,7 @@
 
 command_groups: List[Tuple[str, List[str]]] = [
     ("General topics",
-     ["file", "format", "help", "hooks", "version"]),
+     ["config", "file", "format", "help", "hooks", "version"]),
```
Comment (PawelLipski) [inline git_machete/cli.py:34]: Pls remember to update shell completions

File: git_machete/docs.py [hunk 2]
```diff
@@ -31,6 +32,8 @@
     "version": "Display the version and exit"
 }
 
+help_topics: List[str] = ['config']
```
Comment (PawelLipski) [inline git_machete/docs.py]: Hmmm can it be defined in tests instead? as it's not used anywhere in production code

File: git_machete/cli.py (hunk 1 above)
Comment (PawelLipski) [inline git_machete/cli.py:34]: Pls remember to update shell completions

File: git_machete/docs.py [hunk 3]
```diff
@@ -182,6 +202,36 @@
...
+
+          {textwrap.indent(status_extraSpaceBeforeBranchName_config_key, "          ").strip()}
+
+        - `machete.worktree.useTopLevelMacheteFile`
+
+          # to-do
+
+        <b>Environment variables: </b>
+
+        - `GIT_MACHETE_REBASE_OPTS`
```
Comment (PawelLipski) [inline git_machete/docs.py:233]: Also `GIT_MACHETE_EDITOR` is to be documented... pls grep the project for other env vars but I don't think there are any others

File: completion/git-machete.completion.bash [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 _git_machete() {
   local cmds="add advance anno clean d delete-unmanaged diff discover e edit file fork-point g github go help is-managed l list log reapply s show slide-out squash status traverse update version"
-  local help_topics="$cmds format hooks"
+  local help_topics="$cmds format hooks config"
```
Comment (PawelLipski) [inline completion/git-machete.completion.bash]: Nit:
```suggestion
  local help_topics="$cmds config format hooks"
```

File: completion/git-machete.completion.zsh [hunk 5]
```diff
@@ -192,6 +192,7 @@ __git_machete_help_topics() {
   topics+=(
     'format:Format of the .git/machete definition file'
     'hooks:Display docs for the extra hooks added by git machete'
+    'config:Docs for the configuration keys and environment variables'
```
Comment (PawelLipski) [inline completion/git-machete.completion.zsh]: Nit, move up for abc order

File: completion/git-machete.fish [hunk 6]
```diff
@@ -6,7 +6,7 @@
 # - https://gitter.im/fish-shell/fish-shell?at=6176a4bacd4972068b6a0804
 complete --do-complete="git "
 
-set -l __mcht_help_topics format hooks
+set -l __mcht_help_topics format hooks config
```
Comment (PawelLipski) [inline completion/git-machete.fish]:
```suggestion
set -l __mcht_help_topics config format hooks
```

File: git_machete/docs.py [hunk 7]
```diff
@@ -226,11 +224,15 @@
...
-          Example: `GIT_MACHETE_REBASE_OPTS="--keep-empty --rebase-merges" git machete update`
+          Example: `GIT_MACHETE_REBASE_OPTS="--keep-empty --rebase-merges" git machete update`.
 
         - `GITHUB_TOKEN`
 
           Used to store GitHub API token.
+          
+        - `GIT_MACHETE_EDITOR`
+        
+          Used to store editor name with witch you edit the definition file manually via `git machete e[dit]`.
```
Comment (PawelLipski) [inline git_machete/docs.py]: _witch_? 🧙🏻‍♀️

File: git_machete/docs.py [hunk 8]
```diff
@@ -226,11 +224,15 @@
...
 
           Used to pass extra options to the underlying `git rebase` invocation (called by the executed command, such as: `reapply`, `slide-out`, `traverse`, `update`
-          Example: `GIT_MACHETE_REBASE_OPTS="--keep-empty --rebase-merges" git machete update`
+          Example: `GIT_MACHETE_REBASE_OPTS="--keep-empty --rebase-merges" git machete update`.
 
         - `GITHUB_TOKEN`
 
           Used to store GitHub API token.
+          
+        - `GIT_MACHETE_EDITOR`
```
Comment (PawelLipski) [inline git_machete/docs.py]: Abc order pls

File: tests/test_help.py [hunk 9]
```diff
@@ -27,6 +27,7 @@ def test_help(self, mocker: Any) -> None:
         mocker.patch('git_machete.utils.run_cmd', mock_run_cmd)  # to hide git outputs in tests
 
         expected_exit_code = None
+        help_topics: List[str] = ['config']
```
Comment (PawelLipski) [inline tests/test_help.py]: I'd say that `format` and `hooks` are also _help topics_ (and not commands)... why just `config` is included here? 🤔

File: git_machete/docs.py (hunk 7 above)
Comment (amalota) [reply git_machete/docs.py]: hahahahaha :D

File: tests/test_help.py (hunk 9 above)
Comment (amalota) [reply tests/test_help.py]: Well, it turns out that both `format` and `hooks` are actually commands that do nothing :D

`    subparsers.add_parser(
        'format', add_help=False, usage=argparse.SUPPRESS, parents=[common_args_parser])`

`    subparsers.add_parser(
        'hooks',
        usage=argparse.SUPPRESS,
        add_help=False,
        parents=[common_args_parser])`

@PawelLipski i should remove them, right?

Comment (PawelLipski) [reply tests/test_help.py]: Whooops, yeah pls remove them, ofc making sure that they're still covered by `test_help.py` as help topics

**Comment by PawelLipski** [comment]
Huh coverage decrease? 🤔 that's suspicious!

**Comment by PawelLipski** [comment]
Jesus, Codecov is really pog, thanks @AABur ❤️

**Comment by amalota** [comment]
> Huh coverage decrease? 🤔 that's suspicious!

Hmm, i think it's because of `help_topics` exclusion from the `test_help`

image

**Comment by PawelLipski** [comment]
> Hmm, i think it's because of help_topics exclusion from the test_help

Okay... but I suppose it's not intended?

**Comment by amalota** [comment]
> > Hmm, i think it's because of help_topics exclusion from the test_help
> 
> Okay... but I suppose it's not intended?

No, i had to do it so the tests will pass

**Comment by PawelLipski** [comment]
> No, i had to do it so the tests will pass

Can you fix the tests so that they pass without coverage decrease? 🤔

**Comment by amalota** [comment]
@PawelLipski Hm, there is sth weird going on with codecov, there is no new report for the newest commit because codecov shows that CI for that commit failed (https://app.codecov.io/gh/VirtusLab/git-machete/commits?branch=docs%2F518-help-topic-for-config) although you can clearly see here that it passed :P I don't know how to force codecov to comment the newest report

**Comment by PawelLipski** [comment]
Huh weird... anyway, as long as we don't actually lose coverage (in the sense of certain cases no longer being covered), it's okay

File: git_machete/docs.py [hunk 10]
```diff
@@ -54,6 +55,23 @@
         repository = <repo_name>
         remote = <remote_name>'''
 
+status_extraSpaceBeforeBranchName_config_key = '''To make it easier to select branch name from the `git machete status` output on certain terminals (e.g. Alacritty), you can add an extra
+space between └─ and branch name by setting `git config machete.status.extraSpaceBeforeBranchName true`.
+
+For example, by default it's:
```
Comment (PawelLipski) [inline git_machete/docs.py]:
```suggestion
For example, by default the status is displayed as:
```

File: git_machete/docs.py [hunk 11]
```diff
@@ -182,6 +200,40 @@
           <b>--c, --checkout-my-github-prs</b>     Checkout your open PRs into local branches.
           <b>-y, --yes</b>                         Don't ask for confirmation when deleting branches from git.
     """,
+    "config": f"""
+        <b>Config keys: </b>
+        - `machete.github.(remote,organization,repository)`:
```
Comment (PawelLipski) [inline git_machete/docs.py]:
```suggestion
        - `machete.github.{remote,organization,repository}`:
```

Parens `()` are uncommon for expansion in CLI docs in general; braces `{}` are better suited as they evoke [shell brace expansion](https://virtuslab.github.io/tips/#shell/shell-brace-expansion)

File: git_machete/docs.py [hunk 12]
```diff
@@ -182,6 +200,40 @@
...
+
+          When executing `git machete github <subcommand>` command, {textwrap.indent(github_config_keys, "          ").strip()}
+
+        - `machete.overrideForkPoint.<branch>.(to,whileDescendantOf)`
+
+          Executing `git machete fork-point --override-to=<revision> [<branch>]`, sets up a fork point override for <branch>.
```
Comment (PawelLipski) [inline git_machete/docs.py]:
```suggestion
          Executing `git machete fork-point --override-to=<revision> [<branch>]` sets up a fork point override for <branch>.
```

No need to separate subject and verb (_podmiot od orzeczenia_) by comma in English

File: git_machete/docs.py [hunk 13]
```diff
@@ -182,6 +200,40 @@
...
+          The override data is stored under `machete.overrideForkPoint.<branch>.to` and `machete.overrideForkPoint.<branch>.whileDescendantOf` git config keys.
+
+        - `machete.status.extraSpaceBeforeBranchName`
+
+          {textwrap.indent(status_extraSpaceBeforeBranchName_config_key, "          ").strip()}
+
+        - `machete.worktree.useTopLevelMacheteFile`
+
+          # to-do
```
Comment (PawelLipski) [inline git_machete/docs.py]: Nit: you can even add `TODO` for the actual issue number

File: git_machete/docs.py [hunk 14]
```diff
@@ -182,6 +200,40 @@
...
+
+        <b>Environment variables: </b>
+
+        - `GIT_MACHETE_EDITOR`
+
+          Used to store editor name with which you edit the definition file manually via `git machete e[dit]`.
```
Comment (PawelLipski) [inline git_machete/docs.py]: Too verbose:
```suggestion
          Name of the editor used by `git machete e[dit]`, e.g. `vim` or `nano`.
```

File: git_machete/docs.py [hunk 15]
```diff
@@ -182,6 +200,40 @@
...
+
+        - `GIT_MACHETE_REBASE_OPTS`
+
+          Used to pass extra options to the underlying `git rebase` invocation (called by the executed command, such as: `reapply`, `slide-out`, `traverse`, `update`
```
Comment (PawelLipski) [inline git_machete/docs.py]: Close the paren

File: git_machete/docs.py [hunk 16]
```diff
@@ -182,6 +200,40 @@
...
+          Example: `GIT_MACHETE_REBASE_OPTS="--keep-empty --rebase-merges" git machete update`.
+
+        - `GITHUB_TOKEN`
+
+          Used to store GitHub API token.
```
Comment (PawelLipski) [inline git_machete/docs.py]: Mention what commands use it

File: git_machete/docs.py [hunk 17]
```diff
@@ -678,22 +730,7 @@
...
-            └─feature_branch2
-
-        With `machete.status.extraSpaceBeforeBranchName` config set to `true`:
-            develop
-            │
-            ├─ feature_branch1
-            │
-            └─ feature_branch2
+{textwrap.indent(status_extraSpaceBeforeBranchName_config_key, "          ")}
     """,
```
Comment (PawelLipski) [inline git_machete/docs.py:741]: Pls reflect all changes in .rst once this is complete

**Comment by amalota** [comment]
> Huh weird... anyway, as long as we don't actually lose coverage (in the sense of certain cases no longer being covered), it's okay

@PawelLipski Yes, each case is being covered same as before just differently

File: git_machete/docs.py [hunk 18]
```diff
@@ -202,13 +202,13 @@
...
     "config": f"""
         <b>Config keys: </b>
-        - `machete.github.(remote,organization,repository)`:
+        - `machete.github.{{remote,organization,repository}}`:
```
Comment (PawelLipski) [inline git_machete/docs.py:207]: Huh double brace? 🤔

Comment (amalota) [reply git_machete/docs.py:207]: @PawelLipski Yeah, its an f string so `{` and `}` has to be escaped, i found that you can do it with double brace

Comment (PawelLipski) [reply git_machete/docs.py:207]: Awww crap indeed

**Comment by amalota** [comment]
@PawelLipski config.rst result:

image

image

image

**Comment by PawelLipski** [comment]
> @PawelLipski config.rst result:

gr8! just sort out the double braces

File: docs/source/cli_help/config.rst [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+                remote = <remote_name>
+
+    * ``machete.overrideForkPoint.<branch>.{{to,whileDescendantOf}}``
+
+        Executing ``git machete fork-point --override-to=<revision> [<branch>]`` sets up a fork point override for <branch>.
+        The override data is stored under ``machete.overrideForkPoint.<branch>.to`` and ``machete.overrideForkPoint.<branch>.whileDescendantOf`` git config keys.
+
+    * ``machete.status.extraSpaceBeforeBranchName``
+
+        To make it easier to select branch name from the ``status`` output on certain terminals
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst:40]: Pls state the default (`false` IIRC) explicitly in the first sentence, as you did for `machete.worktree.useTopLevelMacheteFile`

Comment (PawelLipski) [reply docs/source/cli_help/config.rst:40]: (I mean, it is implied, but still)

File: docs/source/cli_help/config.rst [hunk 20]
```diff
@@ -0,0 +1,82 @@
+.. _config:
+
+config
+-----
+Documentation about available ``git machete`` config keys and environment variables that change the command's default behavior.
+
+Note: ``config`` is not a command as such, just a help topic (there is no ``git machete config`` command).
+
+**Config keys:**
+    * ``machete.github.{{remote,organization,repository}}``:
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst]: Yeah, single braces FTW

File: docs/source/cli_help/config.rst [hunk 21]
```diff
@@ -60,23 +60,23 @@ Note: ``config`` is not a command as such, just a help topic (there is no ``git
            │
            └─ feature_branch2
 
-    * ``machete.worktree.useTopLevelMacheteFile``
+    * `machete.worktree.useTopLevelMacheteFile`
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst]: Hmmm single backtick? 🤔 doesn't it mean italics in .rst?

File: docs/source/cli_help/config.rst [hunk 22]
```diff
@@ -7,7 +7,7 @@ Documentation about available ``git machete`` config keys and environment variab
 Note: ``config`` is not a command as such, just a help topic (there is no ``git machete config`` command).
 
 **Config keys:**
-    * ``machete.github.{{remote,organization,repository}}``:
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst]: Single brace! not single backtick

File: docs/source/cli_help/config.rst [hunk 23]
```diff
@@ -0,0 +1,82 @@
+.. _config:
+
+config
+-----
+Documentation about available ``git machete`` config keys and environment variables that change the command's default behavior.
+
+Note: ``config`` is not a command as such, just a help topic (there is no ``git machete config`` command).
+
+**Config keys:**
+    * `machete.github.{{remote,organization,repository}}`:
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst]: Single brace! not single backtick

File: git_machete/docs.py [hunk 24]
```diff
@@ -182,6 +201,42 @@
           <b>--c, --checkout-my-github-prs</b>     Checkout your open PRs into local branches.
           <b>-y, --yes</b>                         Don't ask for confirmation when deleting branches from git.
     """,
+    "config": f"""
+        <b>Config keys: </b>
```
Comment (PawelLipski) [inline git_machete/docs.py:205]: Nit: add extra newline to match `Environment variables:` section below

File: git_machete/docs.py [hunk 25]
```diff
@@ -182,6 +201,42 @@
...
+        - `machete.status.extraSpaceBeforeBranchName`
+
+          {textwrap.indent(status_extraSpaceBeforeBranchName_config_key, "          ").strip()}
+
+        - `machete.worktree.useTopLevelMacheteFile`
+
+          The default value of this key is `true`, which means that the path to machete definition file will be `.git/machete`
+          for both regular directory and worktree. If you want the worktree to have its own machete definition file (located under
+          `.git/worktrees/.../machete`), set `git config machete.worktree.useTopLevelMacheteFile false`.
+
```
Comment (PawelLipski) [inline git_machete/docs.py:225]: Plus some extra newline here to visually separate the two top-level sections

File: docs/source/cli_help/config.rst [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+Documentation about available ``git machete`` config keys and environment variables that change the command's default behavior.
+
+Note: ``config`` is not a command as such, just a help topic (there is no ``git machete config`` command).
+
+**Config keys:**
+
+    * ``machete.github.{remote,organization,repository}``:
+
+        When executing ``git machete github <subcommand>`` command, GitHub API server URL will be inferred from ``git remote``.
+        You can override this by setting the following git config keys:
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst]: I'd emphasise that:

```suggestion
        You can override this by setting the following local git config keys:
```

As setting it globally completely doesn't make sense

File: git_machete/docs.py [hunk 27]
```diff
@@ -41,7 +41,7 @@
     4. current auth token from the `hub` GitHub CLI.'''
 
 github_config_keys = '''GitHub API server URL will be inferred from `git remote`.
-You can override this by setting the following git config keys:
+You can override this by setting the following local git config keys:
```
Comment (PawelLipski) [inline git_machete/docs.py:44]: Pls reflect in .rst

File: docs/source/cli_help/config.rst [hunk 28]
```diff
@@ -67,6 +67,7 @@ Note: ``config`` is not a command as such, just a help topic (there is no ``git
         for both regular directory and worktree. If you want the worktree to have its own machete definition file (located under
         ``.git/worktrees/.../machete``), set ``git config machete.worktree.useTopLevelMacheteFile false``.
```
Comment (PawelLipski) [inline docs/source/cli_help/config.rst:69]: Pls attach a screen how .rst looks after all the changes

Comment (amalota) [reply docs/source/cli_help/config.rst:69]: image

image

**Review by PawelLipski — APPROVED** [review]

**Pull request merged**